cranelift/codegen/src/isa/aarch64/abi.rs

```diff
@@ -215,22 +307,43 @@ fn compute_arg_locs(call_conv: isa::CallConv, params: &[ir::AbiParam]) -> (Vec<A
 
 impl ABISig {
     fn from_func_sig(sig: &ir::Signature) -> ABISig {
-        // Compute args and retvals from signature.
-        // TODO: pass in arg-mode or ret-mode. (Does not matter
-        // for the types of arguments/return values that we support.)
-        let (args, stack_arg_space) = compute_arg_locs(sig.call_conv, &sig.params);
-        let (rets, _) = compute_arg_locs(sig.call_conv, &sig.returns);
-
-        // Verify that there are no return values on the stack.
-        debug_assert!(rets.iter().all(|a| match a {
-            &ABIArg::Stack(..) => false,
-            _ => true,
-        }));
+        // Compute args and retvals from signature. Handle retvals first,
+        // because we may need to add a return-area arg to the args.
+        let (rets, stack_ret_space) = compute_arg_locs(
+            sig.call_conv,
+            &sig.returns,
+            ArgsOrRets::Rets,
+            /* extra ret-area ptr = */ false,
+        );
+        let need_stack_return_area = stack_ret_space > 0;
+        let (args, stack_arg_space) = compute_arg_locs(
+            sig.call_conv,
+            &sig.params,
+            ArgsOrRets::Args,
+            need_stack_return_area,
+        );
+        let stack_ret_arg = if need_stack_return_area {
+            Some(args.len() - 1)
```

Comment:
This is breaking the encapsulation with `compute_arg_locs`, since the index of the stack return area pointer is chosen there; could this function return it, instead?